### PR TITLE
Fix false positives for the codec registration checker

### LINKED ISSUE #141: False-positive `No codecs for ... registered` error when switching branches (sth wrong with incremental compilation?)
Waiting for further details from the team. Just the combination of git commit + contents of `target/codec-registration-checker-cache.csv` should be enough.

### PR DESCRIPTION
**Pull request by LukaszKontowski**
Closes #141

- false positives fixed - both for situations, where git branch is changed in the middle of sbt compilations, and for situations where on the same branch new registrations has been added and later deleted
- as usage of https://github.com/pathikrit/better-files has been added to the runtime logic of the codec-registration-checker-compiler-plugin - this dependency has to be provided within a fat jar when ASH gets published - hence changes inside build.sbt
- description with instructions in README added (one new parameter for codec-registration-checker-compiler-plugin)
- added logic - checking source code directory for files containing possible "positives" - gets invoked only if such possible positives were detected - moreover - works quick (just few seconds on Hydra if there are possible missing registrations). So there is no real negative impact on performance.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -304,6 +304,13 @@ class ExampleSerializer(actorSystem: ExtendedActorSystem)
 Note that as with Serializability Checker and Dump Persistence Schema,
 this compiler plugin only runs in the sbt modules where `AkkaSerializationHelperPlugin` is explicitly enabled.
 
+To make this plugin work correctly, path to the source code directory where codecs are registered has to be set in sbt project settings - just add settings like below:
+```
+.settings(scalacOptions ++=
+    Seq(s"-P:codec-registration-checker-plugin:--source-code-directory=${baseDirectory.value.file.getAbsolutePath}/path/to/your/code/directory"))
```
Comment (PawelLipski) [inline README.md:310]: Huh that'll make the usage pretty complicated 🤔

Could you explain why passing the source code directory is needed to prevent the error?

Comment (PawelLipski) [reply README.md:310]: Also... even if it stays as is, I'd use a more realistic path as a sample (with `src/main/scala` etc.)

File: build.sbt [hunk 2]
```diff
@@ -56,6 +56,26 @@ lazy val commonSettings = Seq(
       if (sys.env.getOrElse("CI", "false") == "true") "-Xfatal-warnings" else ""),
   libraryDependencies ++= commonDeps)
 
+lazy val assemblySettings = Seq(
```
Comment (PawelLipski) [inline build.sbt:59]: Pls document on why that's needed! (plugin compiler classpath etc.)

Comment (PawelLipski) [reply build.sbt:59]: Actually... lemme open a separate task for researching the topic

File: README.md (hunk 1 above)
Comment (LukaszKontowski) [reply README.md:310]: > Huh that'll make the usage pretty complicated 🤔
> 
> Could you explain why passing the source code directory explicitly is needed to prevent the error?

Yes - because the only way to check if these missing types are "false-positives" or not - is by checking the source code in the repo that uses ASH plugins. This check does recursive search on all .scala files in the `--source-code-directory` for these possibly missing types. In order not to check whole repository (e.g. several modules in a bigger project) - it would be beneficial to specify the `--source-code-directory` - to check just the directory, where codecs are actually registered in code.

It might be also possible not to specify `path/to/your/code/directory` and to check whole repository by just checking ${baseDirectory.value.file.getAbsolutePath}
It would be one-step-easier for clients but in such situation - all .scala files from whole repo would be checked.

Unfortunately - this check has to be done on source code. Due to how incremental compilation works in sbt - we would not ba able to collect all "really existing" types using scala reflection.

File: build.sbt (hunk 2 above)
Comment (PawelLipski) [reply build.sbt:59]: https://github.com/VirtusLab/akka-serialization-helper/issues/168

File: README.md (hunk 1 above)
Comment (LukaszKontowski) [reply README.md:310]: alternatively we can just set `${baseDirectory.value.file.getAbsolutePath}/src/main/scala` as a default ?

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala [hunk 3]
```diff
@@ -63,10 +75,12 @@ class CodecRegistrationCheckerCompilerPlugin(override val global: Global) extend
         false
     }
   }
-  override val optionsHelp: Option[String] = Some("""
+  override val optionsHelp: Option[String] = Some(s"""
       |. - directory where cache file will be saved, required
-      |--disable - disables the plugin
+      |$disableFlag - disables the plugin
+      |$sourceCodeDirectoryFlag - path of the source code directory, which has to be checked with this plugin
```
[the remark this answers is not in the document]
Comment (PawelLipski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala:81]: Hmmm `src/main/scala/hydra`, or rather `services/src/main/scala/hydra`?

As top-level `src/main/scala/` does not exist:

image

File: README.md (hunk 1 above)
Comment (PawelLipski) [reply README.md:310]: (see the other comment about Hydra's case, for which this directory will **not** exist IIUC)

Comment (LukaszKontowski) [reply README.md:310]: It will work - see my reply on other comment - ${baseDirectory.value.file.getAbsolutePath} evaluates to the module (subproject) directory where plugin is enabled

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (PawelLipski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala:81]: As a small improvement here, I'd extract the path to Scala source code directory programatically from sbt, shouldn't be too difficult (rather than hardcoding `src/main/scala`), let's then think what to do next

File: README.md (hunk 1 above)
Comment (PawelLipski) [reply README.md:310]: >  as a default ?

Okay... so will there be cases where ones will **not** want to simply set this flag to `src/main/scala` in the current subproject? 🤔

Comment (PawelLipski) [reply README.md:310]: Okay! pls proceed

Comment (MarconZet) [reply README.md:310]: Check out `scalaSource.value`

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala [hunk 4]
```diff
@@ -176,6 +160,69 @@ class SerializerCheckCompilerPluginComponent(
             None
         }
       }
+
+      private def collectMissingClassNames(fullyQualifiedClassNames: List[String]): List[String] = {
```
Comment (MarconZet) [inline codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: If I may provide an alternative solution to this problem.

For this problem I originally thought about reusing already existing compiler phase.

The non-existing FQCN are resolved by typer compiler phase (you can check this by using `-Ystatistics`)

So if we could create our own typer phase (by `global.analyzer.newTyper`) and send it prepared by plugin AST that contains the missing FQCN the phase would resolve these names for us.

Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: Ok, can you specify why adding our own custom typer phase would be a better solution?

Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: My first thought was - it would be smart but it also adds one more phase to each compilation - whereas my solution adds additional method invocation only in situations, where there were some missing types found. In my opinion we do not need one more phase for this particular reason (these false positives should not be that often? rather a rare case?)

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala [hunk 5]
```diff
@@ -24,9 +26,19 @@ class CodecRegistrationCheckerCompilerPlugin(override val global: Global) extend
...
 
   override def init(options: List[String], error: String => Unit): Boolean = {
-    if (options.contains("--disable"))
+    if (options.contains(disableFlag))
       return false
 
+    options.find(flag => flag.contains(sourceCodeDirectoryFlag)) match {
+      case Some(directoryFlag) =>
+        pluginOptions.sourceCodeDirectoryToCheck =
+          directoryFlag.substring(24) // as --source-code-directory= is exactly 24 chars long
```
[the remark this answers is not in the document]
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala]: done

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPlugin.scala:80]: done

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala [hunk 6]
```diff
@@ -176,6 +160,69 @@ class SerializerCheckCompilerPluginComponent(
...
+          val packageName = fqcn.substring(0, indexOfLastDotInFQCN)
+          val typeName = fqcn.substring(indexOfLastDotInFQCN + 1)
+          sourceCodeFilesAsStrings.exists(fileAsString => {
+            fileAsString.startsWith(s"package $packageName") &&
+            (fileAsString.contains(s"class $typeName") || fileAsString.contains(s"trait $typeName") || fileAsString
+              .contains(s"object $typeName"))
+          })
+        }
+
+        for (fqcn <- fullyQualifiedClassNames if typeIsDefinedInScalaFiles(fqcn)) yield fqcn
```
[the remark this answers is not in the document]
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala]: done

File: README.md (hunk 1 above)
Comment (LukaszKontowski) [reply README.md]: I just uploaded working solution - please see https://github.com/VirtusLab/akka-serialization-helper/pull/166/files#diff-d06aaa8abc0aa4f2b1333353fd3962771928ea9140a9ae91b0963b236bf2adceR58

File: build.sbt (hunk 2 above)
Comment (PawelLipski) [inline build.sbt:59]: Actually! pls add a reference to https://github.com/sbt/sbt/issues/2255 + a short explanation in comment here

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CacheFileInteractionMode.scala [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package org.virtuslab.ash
+
+sealed trait CacheFileInteractionMode
+case class DumpTypesIntoCacheFile() extends CacheFileInteractionMode
```
Comment (PawelLipski) [inline codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CacheFileInteractionMode.scala:4]: Nit: just case objects?

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala [hunk 8]
```diff
@@ -176,6 +160,69 @@ class SerializerCheckCompilerPluginComponent(
...
+            fileAsString.startsWith(s"package $packageName") &&
+            (fileAsString.contains(s"class $typeName") || fileAsString.contains(s"trait $typeName") || fileAsString
+              .contains(s"object $typeName"))
+          })
+        }
+
+        fullyQualifiedClassNames.filter(typeIsDefinedInScalaFiles)
+      }
+
+      private def interactWithTheCacheFile(
```
Comment (PawelLipski) [inline codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:183]: Actually, given that this method is only ever called with a specific `mode` known upfront (and not a variable one)... it'll be better to split this method into two independent methods (esp. that the 2 `match` cases are completely separate)

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala [hunk 9]
```diff
@@ -176,6 +160,69 @@ class SerializerCheckCompilerPluginComponent(
...
+          mode: CacheFileInteractionMode,
+          typeNamesToRemove: List[String] = List.empty): Unit =
+        mode match {
+          case _: DumpTypesIntoCacheFile =>
```
Comment (PawelLipski) [inline codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:187]: In a nutshell, this case is the old piece from `def apply`, while the other case is newly implemented, right?

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CacheFileInteractionMode.scala (hunk 7 above)
Comment (PawelLipski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CacheFileInteractionMode.scala:4]: Or better, just get rid of them completely, as suggested in SerializerCheckCompilerPluginComponent.scala

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala (hunk 9 above)
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:187]: yes

File: codec-registration-checker-compiler-plugin/src/test/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPluginSpec.scala [hunk 10]
```diff
@@ -26,7 +28,7 @@ class CodecRegistrationCheckerCompilerPluginSpec extends AnyWordSpecLike with sh
       File.usingTemporaryDirectory() { directory =>
         val out = CodecRegistrationCheckerCompiler.compileCode(
           serializersCode(0) :: dataSourceCode,
-          List(s"${directory.toJava.getAbsolutePath}"))
+          List(s"${directory.toJava.getAbsolutePath}", s"--source-code-directory=$sourceCodeDirectory"))
```
Comment (PawelLipski) [inline codec-registration-checker-compiler-plugin/src/test/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPluginSpec.scala:31]: Nit: pls add a case where `--source-code-directory` is **not** passed and exception is thrown

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala (hunk 4 above)
Comment (MarconZet) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: > method invocation only in situations, where there were some missing types found

This doesn't change regardless of method

Comment (MarconZet) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: > why adding our own custom typer phase would be a better solution?

It doens't use fat jar, better files and also it does not depend on sbt.

File: build.sbt [hunk 11]
```diff
@@ -189,23 +211,6 @@ lazy val dumpPersistenceSchemaCompilerPlugin = (projectMatrix in file("dump-pers
         }
```
Comment (PawelLipski) [inline build.sbt:211]: Actually... would it make sense to mark `scalaPluginDeps21x` as `Provided`? 🤔

Comment (PawelLipski) [reply build.sbt:211]: So as to avoid packing at least `scala.*` classes into the fat jar

Comment (LukaszKontowski) [reply build.sbt:211]: you mean these ones?
image
I can try to change it if you wish

File: build.sbt (hunk 2 above)
Comment (LukaszKontowski) [reply build.sbt:60]: Sure - comment in file added. Short explanation is in the PR description, but here we go again :)
- as usage of https://github.com/pathikrit/better-files has been added to the runtime logic of the codec-registration-checker-compiler-plugin - this dependency has to be provided within a fat jar when ASH gets published - hence changes inside build.sbt
TL;DR: without publishing fat jar our commercial project (Hydra) did not find classes better-files during compilation - that's why fat-jar was needed

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CacheFileInteractionMode.scala (hunk 7 above)
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/CacheFileInteractionMode.scala]: File removed - according to other comment (no need to have these types as interactWithTheCacheFile has been replaced with two separate methods)

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala (hunk 8 above)
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala]: done

File: codec-registration-checker-compiler-plugin/src/test/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPluginSpec.scala (hunk 10 above)
Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/test/scala/org/virtuslab/ash/CodecRegistrationCheckerCompilerPluginSpec.scala:31]: sure, test added

File: build.sbt (hunk 2 above)
Comment (PawelLipski) [reply build.sbt:60]: Yes! I do get it... I meant a comment **in code**, coz locating comments on GitHub via `git blame` isn't particularly convenient :/

Comment (PawelLipski) [reply build.sbt:60]: Assume someone would revisit this place in, say, 3 years 😅

Comment (LukaszKontowski) [reply build.sbt:60]: Ok, so - is the comment in build.sbt enough - or write some more detailed explanation?

File: sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala [hunk 12]
```diff
@@ -51,7 +54,8 @@ object AkkaSerializationHelperPlugin extends AutoPlugin {
         new File(
           (ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputDirectoryPath).value) / (ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputFilename).value,
       ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputFilename := s"${name.value}-dump-persistence-schema-${version.value}.yaml",
-      ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputDirectoryPath := ashCompilerPluginCacheDirectory.value.getPath) ++
+      ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputDirectoryPath := ashCompilerPluginCacheDirectory.value.getPath,
+      ashCodecRegistrationCheckerCompilerPlugin / sourceCodeDirectory := s"--source-code-directory=${baseDirectory.value.getAbsolutePath}${File.separator}src${File.separator}main${File.separator}scala") ++
```
Comment (PawelLipski) [inline sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala:58]: Okay... so just this one potentially left to sort out... other than that, PR ready to merge I think :)

File: codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala (hunk 4 above)
Comment (PawelLipski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: (@LukaszKontowski consider opening a separate _self-contained_ issue, i.e. with all known details outlined &mdash; no reference to this thread, for the sake of clarity)

Comment (PawelLipski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: It's hard to me to assess how significant the suggested improvement is... but I'd intuitively go for 2. as well 😅

Comment (LukaszKontowski) [reply codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala:164]: Sure, I'll create a new issue for this improvement with detailed description.

File: build.sbt (hunk 2 above)
Comment (LukaszKontowski) [reply build.sbt:62]: done

File: sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala (hunk 12 above)
Comment (LukaszKontowski) [reply sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala]: switched to use `sourceDirectory` - unfortunately - did not find more elegant working solution

File: sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala [hunk 13]
```diff
@@ -55,7 +55,7 @@ object AkkaSerializationHelperPlugin extends AutoPlugin {
           (ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputDirectoryPath).value) / (ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputFilename).value,
       ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputFilename := s"${name.value}-dump-persistence-schema-${version.value}.yaml",
       ashDumpPersistenceSchema / ashDumpPersistenceSchemaOutputDirectoryPath := ashCompilerPluginCacheDirectory.value.getPath,
-      ashCodecRegistrationCheckerCompilerPlugin / sourceCodeDirectory := s"--source-code-directory=${baseDirectory.value.getAbsolutePath}${File.separator}src${File.separator}main${File.separator}scala") ++
+      ashCodecRegistrationCheckerCompilerPlugin / sourceCodeDirectory := s"--source-code-directory=${sourceDirectory.value.getAbsolutePath}${File.separator}main${File.separator}scala") ++
```
[the remark this answers is not in the document]
Comment (LukaszKontowski) [reply sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala]: comment added

[the remark this answers is not in the document]
Comment (PawelLipski) [reply sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperPlugin.scala]: Actually! for future, unless your reply is non-trivial, you can just `Resolve` the comment without further reply
